### PR TITLE
BUG/ENH: Fix Hamilton filter docstrings and handle order=0, closes #4248

### LINKED ISSUE #4248: DOC/BUG: `py_hamilton_filter` doesn't handle time-dependent transition probabilities the same as `cy_hamilton_filter`.
The documentation for statsmodels.tsa.regime_switching.markov_switching.py_hamilton_filter suggests that it can handle time-dependent transition probabilities, but the following script fails when I run it.

`
import sys
import numpy as np
import statsmodels
import statsmodels.tsa.regime_switching.markov_switching as ms

print "Python version:", sys.version
print "Numpy version:", np.version.full_version
print "statsmodels version:", statsmodels.version.full_version

np.random.seed(10)
k_regimes = 2
nobs = 4
order = 1

_initial_probabilities = np.arange(1, k_regimes + 1, dtype='float64')
initial_probabilities = _initial_probabilities / np.sum(_initial_probabilities)

_regime_transition = np.random.exponential(size=(k_regimes, k_regimes, nobs))
regime_transition = _regime_transition / np.sum(_regime_transition, axis=0)

conditional_likelihoods_shape = (1 + order) * (k_regimes,) + (nobs,)
conditional_likelihoods = np.random.exponential(size=conditional_likelihoods_shape)

res1 = ms.cy_hamilton_filter(initial_probabilities, regime_transition, conditional_likelihoods)
res2 = ms.py_hamilton_filter(initial_probabilities, regime_transition, conditional_likelihoods)
`

Result on my machine:

C:\Data\python\venv\lib\site-packages\statsmodels\compat\pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools
Python version: 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:53:40) [MSC v.1500 64 bit (AMD64)]
Numpy version: 1.13.3
statsmodels version: 0.8.0
Traceback (most recent call last):
  File "temp/example.py", line 26, in <module>
    res2 = ms.py_hamilton_filter(initial_probabilities, regime_transition, conditional_likelihoods)
  File "C:\Data\python\venv\lib\site-packages\statsmodels\tsa\regime_switching\markov_switching.py", line 198, in py_hamilton_filter
    regime_transition[..., transition_t] *
IndexError: index 3 is out of bounds for axis 2 with size 3

This may just be a documentation bug, as it looks like `regime_transition` needs to be longer than `nobs`, but then it seems like the cython version should also fail above.

### PR DESCRIPTION
**Pull request by gbrunick**
This pull request makes some improvements to the Hamilton filter functionality and docs (both the Python and Cython versions).

* Handles the order=0 case.
* Documents the sizes of the arguments correctly.
* Checks that the argument sizes satisfy the documented sizes.

### CODE REVIEW

**Comment by gbrunick** [comment]
Looks like build #6803.5 failed.  I see:

> Error:  Dependencies missing in current linux-64 channels: ...

I'm not sure where I go from here.

**Comment by ChadFulton** [comment]
Thanks very much for this PR! I will try and have a look as soon as I can (unfortunately these are a very busy couple of weeks for me).

The test failure is unrelated.

Thanks again!

File: statsmodels/tsa/regime_switching/tests/test_markov_regression.py [hunk 1]
```diff
@@ -940,6 +940,106 @@ def test_smoother_output(self, **kwargs):
...
+        nobs = 8
+        order = 3
+
+        initial_probabilities = np.ones(k_regimes) / k_regimes
+        regime_transition = np.ones((k_regimes, k_regimes, nobs)) / k_regimes
+        conditional_likelihoods = np.ones(order * (k_regimes,) + (nobs,))
+
+        for func in [markov_switching.py_hamilton_filter,
+                     markov_switching.cy_hamilton_filter]:
+            try:
```
Comment (ChadFulton) [inline statsmodels/tsa/regime_switching/tests/test_markov_regression.py]: Very small thing, but can you use `assert_raises` here? I think that's how we generally do things (or at least what I usually do).

**Comment by ChadFulton** [comment]
This is fantastic, thank you very much and my apologies for the slow review. I had one small comment that if you have time to change that would be great, but if you're busy now I can merge as is and fix it myself.

File: statsmodels/tsa/regime_switching/tests/test_markov_regression.py (hunk 1 above)
Comment (gbrunick) [reply statsmodels/tsa/regime_switching/tests/test_markov_regression.py]: Sure; that is easy enough.  I will do this in the next day or two.

**Comment by gbrunick** [comment]
We are now using assert_raises.

**Comment by gbrunick** [comment]
Apologies in advance; I am not terribly familiar with the GitHub work flow.

I was able to pull the latest from statsmodels:master into my fork using the github GUI as explained in the first answer [here.](https://stackoverflow.com/questions/20984802/how-can-i-keep-my-fork-in-sync-without-adding-a-separate-remote/21131381#21131381)

This added an additional merge commit to master in my forked repo, but I just reset to the previous commit to get back in sync with this repo.  I then rebased my feature branch onto master leaving me with:

$ git log -5 --oneline
a8c6f94 Use assert_raises
67e7f5a Correct Hamilton filter docstrings
6dcb2b0 Handle order=0 case in Hamilton filter
1309f15 Merge pull request #4492 from josef-pkt/bug_coint_autolag
379e0e7 BUG: coint use autolag in adfuller call, closes #4490

which seems to be what I want.  I assume that I now want to force push my rebased feature branch back to my repo?  If I do this, github will recognize what just happened and connect a8c6f94 to this pull request?  I figure it can't hurt to confirm I am doing the right thing before I force push anything.

**Comment by gbrunick** [comment]
Looks like I also managed to screw up my last commit, so I will fix that too once I confirm the correct steps for the rebase.

**Comment by josef-pkt** [comment]
> I assume that I now want to force push my rebased feature branch back to my fork?

Yes, once it looks correct locally, then we force push to the same PR branch. The PR will then have the updated branch. 
(The only small annoyance is that the commits shift to the end of the comment list, which is more of a problem in large, long lasting PRs with a long list of comments.)

**Comment by gbrunick** [comment]
I believe we are good to go.

**Comment by josef-pkt** [comment]
rebase looks good
(I don't know anything about the content)

**Comment by josef-pkt** [comment]
Header (and merge commit message) needs a commit code BUG/ENH ?

**Comment by gbrunick** [comment]
I could probably argue for BUG or ENH.  If it must be one, I think its BUG.

**Comment by gbrunick** [comment]
I see the pattern.  I would suggest:

BUG/ENH: Fix Hamilton filter docstrings and handle order=0, closes #4248

**Comment by ChadFulton** [comment]
Thanks very much @gbrunick!

**Pull request merged**